File: script/make/protocol/js/RoleProtocol.js
export default class RoleProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 10101: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 10102: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 10103: {
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 10101: {
                // Role
                // 角色ID
                const roleId = view.getBigUint64(offset, false);
                offset = offset + 8;
                // 角色名
                const roleNameLength = view.getUint16(offset, false);
                offset = offset + 2;
                const roleNameArray = new Uint8Array(view.buffer.slice(offset, offset + roleNameLength));
                const roleName = textDecoder.decode(roleNameArray);
                offset = offset + roleNameLength;
                // 性别
                const sex = view.getUint8(offset, false);
                offset = offset + 1;
                // 职业
                const classes = view.getUint8(offset, false);
                offset = offset + 1;
                // 等级
                const level = view.getBigUint64(offset, false);
                offset = offset + 8;
                // object
                const role = {roleId, roleName, sex, classes, level};
                return {role};
            }
            case 10102: {
                // Asset
                // 金币
                const gold = view.getBigUint64(offset, false);
                offset = offset + 8;
                // 银币
                const silver = view.getUint32(offset, false);
                offset = offset + 4;
                // 铜币
                const copper = view.getBigUint64(offset, false);
                offset = offset + 8;
                // 经验
                const exp = view.getBigUint64(offset, false);
                offset = offset + 8;
                // object
                const asset = {gold, silver, copper, exp};
                return {asset};
            }
            case 10103: {
                // Vip
                // 等级
                const vipLevel = view.getUint8(offset, false);
                offset = offset + 1;
                // 经验
                const exp = view.getBigUint64(offset, false);
                offset = offset + 8;
                // 过期时间
                const expireTime = view.getUint32(offset, false);
                offset = offset + 4;
                // object
                const vip = {vipLevel, exp, expireTime};
                return {vip};
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/erl/role_protocol.erl
-module(role_protocol).
-export([decode/2, encode/2]).
-include("role.hrl").
-include("asset.hrl").
-include("vip.hrl").

-spec decode(Protocol :: non_neg_integer(), Binary :: binary()) -> {ok, [integer() | binary() | list()]} | {error, Protocol :: non_neg_integer(), Binary :: binary()}.
decode(10101, _Rest_ = <<_/binary>>) ->
    {ok, {}};

decode(10102, _Rest_ = <<_/binary>>) ->
    {ok, {}};

decode(10103, _Rest_ = <<_/binary>>) ->
    {ok, {}};

decode(Protocol, Binary) ->
    {error, Protocol, Binary}.


-spec encode(Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()) -> {ok, binary()} | {error, Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()}.
encode(10101, #role{role_id = RoleId, role_name = RoleName, sex = Sex, classes = Classes, level = Level}) ->
    Data10101 = <<RoleId:64, (byte_size(RoleName)):16, (RoleName)/binary, Sex:8, Classes:8, Level:64>>,
    {ok, <<(byte_size(Data10101)):16, 10101:16, Data10101/binary>>};

encode(10102, #asset{gold = Gold, silver = Silver, copper = Copper, exp = Exp}) ->
    Data10102 = <<Gold:64, Silver:32, Copper:64, Exp:64>>,
    {ok, <<(byte_size(Data10102)):16, 10102:16, Data10102/binary>>};

encode(10103, #vip{vip_level = VipLevel, exp = Exp, expire_time = ExpireTime}) ->
    Data10103 = <<VipLevel:8, Exp:64, ExpireTime:32>>,
    {ok, <<(byte_size(Data10103)):16, 10103:16, Data10103/binary>>};

encode(Protocol, Data) ->
    {error, Protocol, Data}.

